### ISSUE TITLE
Would be nice to have an email verification and password reset for dbAuth  

### ISSUE THREAD

**Issue by faFrafa** [mevdschee/php-crud-api#870]
Hi,
do you think it would be possible to add the possibility of verifying the user email when a user register together, and possibly also a password reset functionality?

The config would need the following parameters:
- smtp server and port (to send both verification and password reset emails)
- email address column (like the username column parameter)
- default template for the email subject and body

This would not only complete the dbAuth middleware, but would also allow to grant the registration process to specific email domains (eg, only @this.org and @that.org users can register).

Unfortunately I have very little PHP knowledge to do it on my own.
I hope it's doable.

Cheers,
Fausto

**Comment by mevdschee** [comment]
> in the dbAuth middleware, do you think it would be possible to add the email verification and password reset process?

Certainly possible, but quite a lot of code.

**Comment by KoljaL** [comment]
Whyte not use the mail() function of PHP?

**Comment by nik2208** [comment]
I've quite struggled with php mail function, ended up using phpmailer instead

**Comment by nik2208** [comment]
@NorthFred what do u mean? there's already the password endpoint (different from register

**Comment by NorthFred** [comment]
@nik2208 I was referring to the OP's topic of resetting the password in case the user doesn't remember it (e.g. password reset link...). Did you make this work with phpmailer?

**Comment by nik2208** [comment]
actually I meant, instead of mail() as suggested by Kolial I used phpmailer.
I didn use it to send password reset instructions, I just said I had troubles making mail() work.

there much out there explaining how to use phpmailer.
reguarding the implementation, I've created an endpoint which actually send the email and called it from the fronted when needed

**Comment by NorthFred** [comment]
@nik2208 Right, thanks for clarifying. I misunderstood your reply. I'll do some more research on topic for password reset implementation.

**Comment by NorthFred** [comment]
@nik2208 Well, I'm using the 'dbAuth' middleware of this awesome library for login, logout, registration (signup) and password change. What I am missing is a way to let the user reset their password if they don't remember their login details - i.e. they can no longer log in.  A typical approach to resolve this, is having the back-end send a "password reset" link, via which the user can access a form to reset the password.

I'm more of a front-end (Angular) person than a back-end dev, so any help is greatly appreciated!

**Comment by nik2208** [comment]
@NorthFred have a look [here](https://github.com/bartosz-io/budget-node/tree/94959117e718ab498e5e720769b2066d082e17c0). It's actually pretty advanced angular.
I've taken it as an example to implement my auth provider (using php-crud-api with dbAuth authentication) redirecting to my apps on successful login (using php-crud-api in jwt auth mode).

**Comment by aleppax** [comment]
This is an enhancement that I needed, but there are some considerations about using phpmailer and doubling the filesize of api.php.
I've tried to do it anyway and requiring the files inside the namespace. Don't know if it is correct, I'm not experienced.
#1006

**Comment by apps-caraga** [comment]
I know this is an old topic, just sharing some thoughts on this. Essentially, the PHP-CRUD-API works as it is. Now for emailing, integrating phpmailer seems to add unnecessary bloat and the library would become a REST + Email API. 

Anyway, my idea is for the library to have some kind of event that can trigger async actions, thus enabling suppport for event-driven architecture. 
For example, upon successful registration, a USER_CREATED event will be triggered and an action such as sending activation email can be triggered and handler by a separate email sending API.
